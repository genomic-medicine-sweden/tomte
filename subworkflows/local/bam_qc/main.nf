//
// BAM QC
//

include { PICARD_COLLECTRNASEQMETRICS     } from '../../../modules/nf-core/picard/collectrnaseqmetrics/main'
include { PICARD_COLLECTINSERTSIZEMETRICS } from '../../../modules/nf-core/picard/collectinsertsizemetrics/main'
include { SAMTOOLS_VIEW                   } from '../../../modules/nf-core/samtools/view/main.nf'
include { SAMTOOLS_CONVERT } from '../../../modules/nf-core/samtools/convert/main.nf'
include { SAMTOOLS_INDEX } from '../../../modules/nf-core/samtools/index/main.nf'

workflow BAM_QC {
    take:
    ch_bam            // channel: [mandatory] [ val(meta), path(bam) ]
    ch_fasta          // channel: [mandatory] [ val(meta), path(fasta) ]
    ch_refflat        // channel: [mandatory] [ path(refflat) ]
    ch_rrna_intervals // channel: [mandatory] [ path(intervals) ]

    main:
    ch_versions = Channel.empty()

    // FIXME: Insert pre-processing here using samtools

    SAMTOOLS_INDEX(ch_bam)

    ch_bam_bai = ch_bam.mix(SAMTOOLS_INDEX.out.bai)

    ch_bam_bai.view { it -> "ch_bam_bai ${it}" }

    SAMTOOLS_VIEW(ch_bam_bai, [], [], 'bai')

    PICARD_COLLECTRNASEQMETRICS(
        SAMTOOLS_VIEW.out.bam,
        ch_refflat,
        ch_fasta.map{ _meta, fasta -> fasta },
        ch_rrna_intervals
    )

    PICARD_COLLECTINSERTSIZEMETRICS(
        ch_bam
    )

    ch_versions = ch_versions.mix(PICARD_COLLECTRNASEQMETRICS.out.versions.first())
    ch_versions = ch_versions.mix(PICARD_COLLECTINSERTSIZEMETRICS.out.versions.first())

    emit:
    metrics_general_rna = PICARD_COLLECTRNASEQMETRICS.out.metrics     // channel: [  val(meta), path(txt) ]
    metrics_insert_size = PICARD_COLLECTINSERTSIZEMETRICS.out.metrics // channel: [  val(meta), path(txt) ]
    versions = ch_versions                                            // channel: [ path(versions.yml) ]
}
