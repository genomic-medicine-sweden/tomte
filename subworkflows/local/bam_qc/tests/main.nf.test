nextflow_workflow {

    name "Test Workflow BAM_QC"
    script "../main.nf"
    workflow "BAM_QC"
    config "./nextflow.config"
    tag "bam_qc"

    setup {
        run("GUNZIP", alias: "GUNZIP_FASTA") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'fasta' ],
                    file("${projectDir}/test_data/grch37_chr21.fa.gz", checkIfExists: true)
                ])
                """
            }
        }

        run("GUNZIP", alias: "GUNZIP_GTF") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'gtf' ],
                    file("${projectDir}/test_data/grch37_chr21.gtf.gz", checkIfExists: true)
                ])
                """
            }
        }

        run("GTFTOGENEPRED_REFFLAT", alias: "GTF_TO_REFFLAT") {
            script "../../../../modules/local/gtftorefflat"
            process {
                """
                input[0] = GUNZIP_GTF.out.gunzip.collect()
                """
            }
        }

        run("GET_RRNA_TRANSCRIPTS") {
            script "../../../../modules/local/get_rrna_transcripts"
            process {
                """
                input[0] = GUNZIP_GTF.out.gunzip.collect()
                """
            }
        }

        run("GATK4_CREATESEQUENCEDICTIONARY", alias: "BUILD_DICT") {
            script "../../../../modules/nf-core/gatk4/createsequencedictionary/main.nf"
            process {
                """
                input[0] = GUNZIP_FASTA.out.gunzip
                """
            }
        }

        run("GATK4_BEDTOINTERVALLIST", alias:"BEDTOINTERVALLIST") {
            script "../../../../modules/nf-core/gatk4/bedtointervallist/main"
            process {
                """
                input[0] = GET_RRNA_TRANSCRIPTS.out.bed.map { it -> [ [id:it.name], it ] }
                input[1] = BUILD_DICT.out.dict
                """
            }
        }
    }


    test("Should run without failures") {

        when {
            workflow {
                """
                ch_bam = Channel.of([
                        [ case:'finequagga', id:'ACC5963A1', sample:'ACC5963A1', strandedness:'reverse', sex:'NA', single_end:false, fq_pairs:1, is_fastq:false ],
                        file("${projectDir}/test_data/ACC5963A1.bam", checkIfExists: true)
                    ])
                input[0] = ch_bam
                input[1] = GUNZIP_FASTA.out.gunzip.collect()
                input[2] = GTF_TO_REFFLAT.out.refflat.collect()
                input[3] = BEDTOINTERVALLIST.out.interval_list.map{ meta, interv -> interv }.collect()
                """
            }
        }

        then {
            { assert workflow.success }
            { assert snapshot(
                file(workflow.out.metrics_general_rna.get(0).get(1)).readLines()[4..10],
                file(workflow.out.metrics_insert_size.get(0).get(1)).readLines()[4..10],
                workflow.out.versions
            ).match()}
        }

    }

}
