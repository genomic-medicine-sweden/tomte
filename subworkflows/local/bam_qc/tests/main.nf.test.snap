{
    "Should run without failures": {
        "content": [
            [
                "",
                "## METRICS CLASS\tpicard.analysis.RnaSeqMetrics",
                "PF_BASES\tPF_ALIGNED_BASES\tRIBOSOMAL_BASES\tCODING_BASES\tUTR_BASES\tINTRONIC_BASES\tINTERGENIC_BASES\tIGNORED_READS\tCORRECT_STRAND_READS\tINCORRECT_STRAND_READS\tNUM_R1_TRANSCRIPT_STRAND_READS\tNUM_R2_TRANSCRIPT_STRAND_READS\tNUM_UNEXPLAINED_READS\tPCT_R1_TRANSCRIPT_STRAND_READS\tPCT_R2_TRANSCRIPT_STRAND_READS\tPCT_RIBOSOMAL_BASES\tPCT_CODING_BASES\tPCT_UTR_BASES\tPCT_INTRONIC_BASES\tPCT_INTERGENIC_BASES\tPCT_MRNA_BASES\tPCT_USABLE_BASES\tPCT_CORRECT_STRAND_READS\tMEDIAN_CV_COVERAGE\tMEDIAN_5PRIME_BIAS\tMEDIAN_3PRIME_BIAS\tMEDIAN_5PRIME_TO_3PRIME_BIAS\tSAMPLE\tLIBRARY\tREAD_GROUP",
                "28847684\t28757338\t0\t11864324\t11589633\t3183988\t2119393\t0\t99381\t932\t439\t49414\t517\t0.008806\t0.991194\t0\t0.412567\t0.403015\t0.110719\t0.073699\t0.815582\t0.813027\t0.990709\t0.746386\t0.32766\t0.446037\t0.2773\t\t\t",
                "",
                "## HISTOGRAM\tjava.lang.Integer",
                "normalized_position\tAll_Reads.normalized_coverage"
            ],
            [
                "",
                "## METRICS CLASS\tpicard.analysis.InsertSizeMetrics",
                "MEDIAN_INSERT_SIZE\tMODE_INSERT_SIZE\tMEDIAN_ABSOLUTE_DEVIATION\tMIN_INSERT_SIZE\tMAX_INSERT_SIZE\tMEAN_INSERT_SIZE\tSTANDARD_DEVIATION\tREAD_PAIRS\tPAIR_ORIENTATION\tWIDTH_OF_10_PERCENT\tWIDTH_OF_20_PERCENT\tWIDTH_OF_30_PERCENT\tWIDTH_OF_40_PERCENT\tWIDTH_OF_50_PERCENT\tWIDTH_OF_60_PERCENT\tWIDTH_OF_70_PERCENT\tWIDTH_OF_80_PERCENT\tWIDTH_OF_90_PERCENT\tWIDTH_OF_95_PERCENT\tWIDTH_OF_99_PERCENT\tSAMPLE\tLIBRARY\tREAD_GROUP",
                "170\t112\t53\t52\t665553\t171.760746\t86.703244\t104178\tFR\t25\t47\t67\t87\t107\t127\t269\t2301\t7805\t14105\t57899\t\t\t",
                "",
                "## HISTOGRAM\tjava.lang.Integer",
                "insert_size\tAll_Reads.fr_count"
            ],
            [
                "versions.yml:md5,525ff00c2ac6befe143f16b1d4b0bba6",
                "versions.yml:md5,f2ccf683968d7657ac00eaf5cfaf4b7a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.5"
        },
        "timestamp": "2025-09-19T11:25:28.312288194"
    }
}